rj_gameplay/stp/role/assignment/naive.py

```diff
@@ -132,16 +161,20 @@ def assign_prioritized_roles(
             request_idx: int = request_ind[assignment_idx]
             robot_idx: int = robot_ind[assignment_idx]
             cost: float = robot_costs[robot_idx, request_idx]
-
             # If cost is equal to INVALID_COST, then it means that role assignment
             # failed. Throw an exception.
             if cost == INVALID_COST:
                 raise RuntimeError("Hard constraints not satisfied!")
 
-            robot: stp.rc.Robot = free_robots[robot_idx]
-            role_id = keys_list[request_idx]
+            # If the row index is not the unassigned row, fill the request
+            if robot_idx < unassigned_idx:
+                robot: stp.rc.Robot = free_robots[robot_idx]
+                role_id = keys_list[request_idx]
+                flat_results[role_id].assign(robot, cost)
 
-            flat_results[role_id].assign(robot, cost)
+        # If unassigned index chosen, remove it from the list of row indices, which will be removed from free robots
```

Comment:
From what I understand, `free_robots` is the list of unassigned robots. 180 updates the free_robots list by removing any robots that were given a role above. These lines make sure that updating is correct by treating the robots with an unassigned cost as assigned, but with that "unassigned cost." Correct me if I'm wrong @HussainGynai

Reply:
Yeah so basically this is implemented so that the last index of free_robots is the unassigned slot, so basically we don't want to remove this from free robots, since multiple roles can be unassigned. So the line below basically says if the unassigned index is in the index of chosen robots remove it from the list of chosen robots, so when the we remove the chosen robots from the list of free robots, we do not remove the unassigned possibility as well.